gogol/src/Network/Google/Auth.hs

```diff
@@ -110,6 +128,33 @@ initStore :: (MonadIO m, MonadCatch m, AllowScopes s)
           -> m (Store s)
 initStore c l m = exchange c l m >>= fmap Store . liftIO . newMVar
 
+-- | Refresh an 'OAuthToken' if it was expired.
+-- Otherwise, keep original one.
+refreshStore :: (MonadIO m, MonadCatch m, AllowScopes s)
+             => Store s
+             -> Logger
+             -> Manager
+             -> m (Store s)
+refreshStore ss@(Store s) l m = do
+    x  <- retrieveAuthFromStore ss
+    mx <- validate x
+    if mx
+        then pure ss
+        else liftIO . modifyMVar s $ \y -> do
```

Comment:
Since `MVar`s are actually lazy I'm not entirely sure (and could be wrong) but the result of this operation won't be evaluated until the store is read? Some details [here](http://blog.ezyang.com/2011/02/all-about-mvars/) and [here](http://hackage.haskell.org/package/strict-concurrency).